### DISCUSSION TITLE
Renaming folder possible

### DISCUSSION THREAD

**Discussion by 404erreur** [jellyfin/jellyfin#11822] [General] [answered]
Hello all,

I would like to k ow if it's possible to rename a system folder that  contain an library, without rescan this one.
So, the library system path will change.²

ex: stop jellyfin, rename folder and make a sed inside some files to change the new path of library

Many thanks in advance for your help :)

Regards
404

**Comment by Shadowghost** [comment]
This is not possible right now.

**Reply by 404erreur** [reply]
Thanks for your reply shadowghost :)
Do you know if it will be possible later ?

Regards
404

**Reply by Shadowghost** [reply]
Not in the foreseeable future, it'd require significant changes to the database and scanner to make it work universally.

**Reply by 404erreur** [reply]
Many thanks for your reply, that's not that I expected, but I understand the reason.

Many thanks for this great tool too, jellyfin is very great :)

Regards
404

**Comment by TheMelmacian** [comment]
If you use nfo files to save metadata and saved all images (covers, backdrops, logos etc.) alongside your media, a change of the library path shouldn't be a big problem. Yes you need a rescan, but all metadata should be picked up from the nfo files and jellyfin should use the pictures that are already present in the library. Im not sure if the watch status is also saved in the nfo files since it can be differ between users.
- rename the folder
- update library inside jellyfin with the new path
- rescan


It's not trivial and only recommended if you really know what you are doing but there is way to rename the folder without a rescan.
For that you have to update the paths in the database and all config files (collections etc.) manually. You should make a backup of the database and config folders before you do anything of this.

**Reply by 404erreur** [reply]
Many thanks for the tips

**Reply by TheMelmacian** [reply]
During the reorganization of my library I renamed one of the folders and thought I should try the "hard way" (for "scientific purposes") and updated the database and config files instead of a re-scan. As far as I can see everything works as expected.
If you (or someone else) really want to go this route I can provide a small tutorial.

**Reply by 404erreur** [reply]
hello @TheMelmacian 

Please, yes an how to will be great :)

Regards
404

**Reply by TheMelmacian** [reply, marked as answer]
How to (theoretically) rename (or move) a library folders without a re-scan.

**IMPORTANT:** I ran this only once for testing purposes and the result was not 100% error free. You do this on your own risk. I'm not responsible if you damage your jellyfin installation in an irreparable way. So be warned. Make and keep a backup of you config folder so you're able to restore the old state if something goes wrong.

**preperations:**
- on windows machines I recommend "[Agent Ransack](https://www.mythicsoft.com/agentransack/)" to search for files inside a folder, since the windows search is shit
- to edit multiple files at once (collection.xml) I recommend [Notepad++](https://notepad-plus-plus.org/) or something similar that allows to search and replace in all open files at once
- a good and free database tool is [DBeaver](https://dbeaver.io/)
- use your own paths and folder names instead of the placeholders
- use `\` on windows, `/` on Linux (Docker) to split path parts

---

1. shut down jellyfin
2. rename/move the library folder
3. go to your jellyfin config folder
4. make a backup of this folder so you can restore it if something goes wrong
5. search and update the config files:
    1. `config/data/collections`:
        - search for all *.xml files inside that folder and all subfolders that contain the `path/to/oldFolder`
        - replace all occurrences with `path/to/newFolder`
    2. `config/root/default`:
        - open the folder that is named after your library in which you want to rename or move a folder
          if you want to make sure that you pick the right folder search for files containing the `path/to/oldFolder`
        - inside the folder search for the _mblink_ file that contains the `path/to/oldFolder` (usually named `oldFolder.mblink` and open it
        - replace `path/to/oldFolder` with `path/to/newFolder`
        - rename the oldFolder.mblink to newFolder.mblink
          if a mblink with the same name already exists add a `1` at the end (or one number higher than the highest existing number with that filename): `newFolder1.mblink`
        - replace `path/to/oldFolder` with `path/to/newFolder` inside the options.xml
    3. replace `path/to/oldFolder` with `path/to/newFolder` inside all playlists (`config/data/playlists`)
    4. search for all other files inside the config folder that contains the old path
        IMPORTANT: Don't update any database files (`*.db`, `*.db.bak1`, `*.db-wal`, `*.db-shm` etc.) with a text editor since these have to be updated via a database tool.
        Also don't update files that show random text, control characters, NUL etc. inside the text editor. They are binary files and the result of changing them without knowing what they do can be unpredictable.
        `*.log` files can be ignored
6. Database update:
    - open `config/data/library.db` and update all columns that contain the old folder path:
```SQL
UPDATE TypedBaseItems
SET Path = replace(Path, 'path/to/oldFolder', 'path/to/newFolder')
WHERE Path IS NOT NULL;

UPDATE TypedBaseItems
SET Images = replace(Images, 'path/to/oldFolder', 'path/to/newFolder')
WHERE Images IS NOT NULL;

UPDATE TypedBaseItems
SET data = replace(data, 'path/to/oldFolder', 'path/to/newFolder')
WHERE data IS NOT NULL;
```
7. commit changes
8. run `PRAGMA integrity_check` to make sure the database is not corrupted (very unlikely but happened to me the first time I run the updates)
    if the result is not **OK**, close the connection and replace the `config/data/library.db` with the one from the backup and repeat from step 6
9. close the connection to the database
10. restart jellyfin
11. check if everything works as before

---

**Problems I encountered:**
- I had to trigger a rescan for some multi version Videos because only one file was visible on the detail page although the poster showed the number of versions correctly
- my collection library was broken (saw a few error messages in the log when I tried to open the collection library)

It's possible that some entries get corrupted during the update of the data column which is of type BLOB and the `replace` function is intended for TEXT columns. This would explain both errors since they use the data column to determine the alternate versions and the elements of a boxset.
So to make this work, the update script for the data column has to be improved or there is another problem I missed.

If someone tests this please share your results.

---

Restoring the old state:
1. stop jellyfin
2. replace the jellyfin config folder with the backup from step 4
3. rename folder to old name / move to old location
4. restart jellyfin

**Reply by oddstr13** [reply]
I was going to say that it is *possible*, maybe even *feasible*, but *not easily doable*, and here you are outlining at least some of the work that needs to be done.

*Not recommended*, but possible to do with *significant* effort, and at least some foundational understanding of programming and database management.

**Reply by 404erreur** [reply]
Many many thanks :)
